lib/semantics/resolve-names.cpp

```diff
@@ -6051,6 +6062,53 @@ bool OmpAttributeVisitor::Pre(const parser::OpenMPLoopConstruct &x) {
   return true;
 }
 
+const parser::Name &OmpAttributeVisitor::GetLoopIndex(
+    const parser::DoConstruct &x) {
+  auto &loopControl{x.GetLoopControl().value()};
+  using Bounds = parser::LoopControl::Bounds;
+  const Bounds &bounds{std::get<Bounds>(loopControl.u)};
+  return bounds.name.thing;
+}
+
+void OmpAttributeVisitor::ResolveSeqLoopIndexInParallelOrTaskConstruct(
+    const parser::Name &iv) {
+  Symbol *symbol{nullptr};
+  std::vector<OmpContext>::reverse_iterator targetIt;
+  for (auto it{ompContext_.rbegin()}; it != ompContext_.rend(); ++it) {
+    if (parallelSet.test(it->directive) ||
+        taskGeneratingSet.test(it->directive)) {
+      symbol = ResolveOmp(iv, Symbol::Flag::OmpPrivate, it->scope);
+      targetIt = it;
+      break;
+    }
+  }
+  if (symbol) {
+    targetIt++;
+    symbol->set(Symbol::Flag::OmpPreDetermined);
+    iv.symbol = symbol;  // adjust the symbol within region
+    for (auto it{ompContext_.rbegin()}; it != targetIt; ++it) {
+      AddToContextObjectWithDSA(*symbol, Symbol::Flag::OmpPrivate, *it);
+    }
+  }
+}
```

Comment:
Don't you also need to fix all `parse::Name` symbols that were visited before it was determined their symbols is actually private ?
You may need a second pass here on all the statements inside the inner parallel or task construct. If so, I am not sure though what would be the best/easier between doing so every-time you need to fix a symbol, or make two passes overall, one to determine what symbol needs to be fixed in which scope, and another one later to fix all the relate `parser::Name` symbols.

Reply:
Yes, it's a problem & I've been thinking about the solutions since last week. Similar to the comment you made for `omp-symbol08.f90`, this is the issue listed as TODO item a) in this PR description. Two potential solutions:
   * two passes (then I need to make sure that each OpenMP msg is emitted only for once)
   * maintain a list or map (or some other DSes) for all the `Names` processed and go back to fix those

I think that I lean towards the second one.

Reply:
Sorry, missed the TODO in PR description. The map/list/vector/... of `Name`s may be an interesting option indeed.